sycl/include/CL/sycl/device_selector.hpp

```diff
@@ -87,5 +87,23 @@ class __SYCL_EXPORT host_selector : public device_selector {
   int operator()(const device &dev) const override;
 };
 
+namespace ONEAPI {
+namespace detail {
+struct filter;
+} // namespace detail
+
+class __SYCL_EXPORT filter_selector : public device_selector {
+public:
+  filter_selector(std::string filter);
+  int operator()(const device &dev) const override;
+
+private:
+  mutable std::vector<std::shared_ptr<detail::filter>> mFilters;
```

Comment:
Can we store `detail::filter` as a plain value instead of having an `std::shared_ptr` on it?
Note, sharing this value between separate values of `filter_selector` can lead to races because `filter`s values are modified.